logendpoint.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "logendpoint.h"
+
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "mainloop.h"
+#include "util.h"
+
+static bool _logging_start_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_start_timeout();
+}
+
+static bool _alive_timeout_cb(void *data)
+{
+    LogEndpoint *log = static_cast<LogEndpoint *>(data);
+    return log->_alive_timeout();
+}
+
+void LogEndpoint::_send_msg(const mavlink_message_t *msg, int target_sysid)
+{
+    uint8_t data[MAVLINK_MAX_PACKET_LEN];
+    struct buffer buffer { 0, data };
+
+    buffer.len = mavlink_msg_to_send_buffer(data, msg);
+    _mainloop->route_msg(&buffer, target_sysid, msg->sysid);
+
+    _stat.read.total++;
+    _stat.read.handled++;
+    _stat.read.handled_bytes += msg->len;
+    if (msg->magic == MAVLINK_STX)
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink2_header);
+    else
+        _stat.read.handled_bytes += sizeof(struct mavlink_router_mavlink1_header);
+}
+
+char *LogEndpoint::_get_filename(const char *extension)
+{
+    char filename[PATH_MAX] = { };
+
+    time_t t = time(NULL);
+    struct tm *timeinfo = localtime(&t);
+    int r;
+    uint16_t i;
+
+    for (i = 0; i < UINT16_MAX; i++) {
+        if (i) {
+            r = snprintf(filename, sizeof(filename), "%s/%i-%02i-%02i_%02i-%02i-%02i_%u.%s", _logs_dir,
+                         timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday,
+                         timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, i, extension);
+        } else {
+            r = snprintf(filename, sizeof(filename), "%s/%i-%02i-%02i_%02i-%02i-%02i.%s", _logs_dir,
+                         timeinfo->tm_year + 1900, timeinfo->tm_mon + 1, timeinfo->tm_mday,
+                         timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, extension);
+        }
+
+        if (r < 1 || (size_t)r >= sizeof(filename)) {
+            log_error_errno(errno, "Error formatting Log file name: (%m)");
+            return nullptr;
+        }
+
+        if (access(filename, F_OK)) {
```

Comment:
So you're going to a lot of trouble to get a unique filename.  You still have a TOCTOU issue here.  Instead of `access(...)` you would ususally have a `_get_fh(..)` method and use open(..., O_EXCL) to ensure you have an open filehandle to a file you created.